### PR TITLE
Pool `BindingDiagnosticBag`s

### LINKED ISSUE #67785: BindingDiagnosticBags should be pooled
An enormous number of these bags are created while binding.  Leading to around 1% of all allocs in typing scenarios:

image

Pooling these would help reduce the amount of work spent in the GC (currently at least 15% of process time in the server):

image

--

Alternatively, looking at the type, it seems like it might actually benefit from being a ref-struct** that just lives on the stack where needed and which is passed around by-ref.  Currently it is heap allocated so taht it can participate in inheritance (with C# and VB subclasses).  However, neither of hte subclasses actually add any state.  They just add typesafe helper methods for their corresponding languages.  These helpers could be added as extensions in the C#/VB compiler layers.  

** A ref-struct might not work out in practice though due to this type being used in VB.  I don't know where we landed with that.  If ref-structs are not viable, then a normal struct would work as well, we'd just have to be careful with it.

I also see that there's at least one place where this is stored as state (in DecodeWellKnownAttributeArguments).  We'd likely want to change tthat to keep this diagnostics bag something only kept on the stack like everywhere else in the compiler.

### PR DESCRIPTION
**Pull request by jjonescz**
Resolves #67785. Haven't yet figured out how to configure <samp>PerfView.exe</samp> to show me numbers like in that issue to compare effect of this PR.

@CyrusNajmabadi FYI

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/BindingDiagnosticBag.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/BindingDiagnosticBag.cs:1]: Should we be doing this for VB as well?

File: src/Compilers/CSharp/Portable/Binder/BindingDiagnosticBag.cs [hunk 1]
```diff
@@ -14,14 +14,16 @@ namespace Microsoft.CodeAnalysis.CSharp
...
         public static readonly BindingDiagnosticBag Discarded = new BindingDiagnosticBag(null, null);
 
-        public BindingDiagnosticBag()
-            : this(usePool: false)
-        { }
+        private readonly ObjectPool<BindingDiagnosticBag>? _pool;
 
-        private BindingDiagnosticBag(bool usePool)
-            : base(usePool)
+        public BindingDiagnosticBag()
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/BindingDiagnosticBag.cs:25]: Can we revert the accessibility to `private`?

Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/BindingDiagnosticBag.cs:25]: There aren't many usages of the public constructor. Do you know why those were not converted those to `GetInstance`+`Free`?

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 2)

**Comment by AlekseyTs** [comment]
Not sure what is wrong with this PR, but CodeFlow crashes trying to load it.

**Comment by jcouv** [comment]
> Not sure what is wrong with this PR, but CodeFlow crashes trying to load it.

Yes, I noticed that too. Fortunately, the PR is small enough that reviewing on GitHub isn't too bad.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 2)

**Comment by AlekseyTs** [comment]
Merged as #68108.

**Pull request closed without merging**